File: examples_android/app/src/main/java/com/confirmit/testsdkapp/viewmodels/TriggerListViewModel.kt
package com.confirmit.testsdkapp.viewmodels

import androidx.fragment.app.FragmentActivity
import com.confirmit.mobilesdk.TriggerSDK
import com.confirmit.mobilesdk.database.externals.Program
import com.confirmit.mobilesdk.database.externals.Server
import com.confirmit.testsdkapp.AppConfigs
import com.confirmit.testsdkapp.domain.TriggerManager
import com.confirmit.testsdkapp.utils.ServerManager
import com.confirmit.testsdkapp.utils.ServerPref
import com.confirmit.testsdkapp.viewmodels.models.ProgramModel
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.async

class ProgramTriggerResult(var success: Boolean, var status: Boolean)

class TriggerListViewModel(private val context: FragmentActivity) {
    private val triggerManager: TriggerManager = TriggerManager()
    var programs: MutableList<ProgramModel> = mutableListOf()
    var selectedProgram: ProgramModel? = null

    init {
        try {
            for (program in triggerManager.getPrograms()) {
                programs.add(ProgramModel(program))
            }
        } catch (_: Exception) {
        }
    }

    fun reloadPrograms() {
        val exists = mutableListOf<String>()
        val curPrograms = triggerManager.getPrograms()
        for (program in curPrograms) {
            exists.add("${program.serverId}_${program.programKey}")
        }

        programs.removeAll {
            val key = "${it.program.serverId}_${it.program.programKey}"
            !exists.contains(key)
        }

        processProgram(curPrograms)
    }

    fun downloadAllProgramsAsync(completion: () -> Unit) = GlobalScope.async {
        TriggerSDK.testMode = AppConfigs.triggerTestMode
        val updatedSurveys = mutableListOf<Program>()
        for (program in programs) {
            TriggerSDK.downloadAsync(program.program.serverId, program.program.programKey).await()
            try {
                val programs = TriggerSDK.getPrograms(program.program.serverId)
                updatedSurveys.addAll(programs)
            } catch (_: Exception) {
            }
        }
        processProgram(updatedSurveys)
        completion()
    }

    fun downloadProgramAsync(serverId: String, programKey: String, completion: (Boolean) -> Unit) = GlobalScope.async {
        TriggerSDK.testMode = AppConfigs.triggerTestMode
        TriggerSDK.downloadAsync(serverId, programKey).await()
        try {
            processProgram(TriggerSDK.getPrograms(serverId))
            completion(true)
        } catch (_: Exception) {
        }
    }

    fun setNewServer(serverId: String, clientId: String, clientSecret: String): Server {
        val server =  ServerManager.setNewConfiguration(serverId, clientId, clientSecret)
        ServerManager.writeServerToPref(serverId, ServerPref(clientId, clientSecret))
        return server
    }

    private fun processProgram(updatePrograms: List<Program>) {
        for (program in updatePrograms) {
            try {
                val exists = programs.indexOfFirst { it.program.serverId == program.serverId && it.program.programKey == program.programKey }
                if (exists >= 0) {
                    programs[exists].update(program)
                } else {
                    programs.add(ProgramModel(program))
                }
            } catch (_: Exception) {
            }
        }
    }
}

File: examples_ios/MyApp/Sources/Controllers/ViewModels/TriggerListViewModel.swift
import ConfirmitMobileSDK
import Foundation

struct ProgramTriggerResult {
  var success: Bool
  var status: Bool
}

class TriggerListViewModel {
  private let feedbackManager: TriggerManager
  var programs: [ProgramModel]
  var selectedProgram: ProgramModel?

  init() {
    feedbackManager = TriggerManager()
    programs = []
    do {
      for program in feedbackManager.getPrograms() {
        programs.append(try ProgramModel(program: program))
      }
    } catch {}
  }

  func reloadPrograms() {
    var exists = [String]()
    let curPrograms = feedbackManager.getPrograms()
    for program in curPrograms {
      exists.append("\(program.serverId)_\(program.programKey)")
    }

    programs.removeAll {
      let key = "\($0.program.serverId)_\($0.program.programKey)"
      return !exists.contains(key)
    }

    // update
    processProgram(updatePrograms: curPrograms)
  }

  func downloadAllProgram(completion: @escaping () -> Void) {
    let dispatchGroup = DispatchGroup()
    for program in programs {
      dispatchGroup.enter()
      TriggerSDK.download(serverId: program.program.serverId, programKey: program.program.programKey) { _, _ in
        do {
          let program = try TriggerSDK.getPrograms(serverId: program.program.serverId)
          self.processProgram(updatePrograms: program)
        } catch {}
        dispatchGroup.leave()
      }
    }
    dispatchGroup.notify(queue: .main, work: DispatchWorkItem(block: {
      completion()
        }))
  }

  func downloadProgram(serverId: String, programKey: String, completion: @escaping () -> Void) {
    TriggerSDK.download(serverId: serverId, programKey: programKey) { _, _ in
      do {
        let program = try TriggerSDK.getPrograms(serverId: serverId)
        self.processProgram(updatePrograms: program)
      } catch {}
      completion()
    }
  }

  func getServerName(serverId: String) -> String {
    do {
      guard let server = try ConfirmitServer.getServer(serverId: serverId) else {
        return ""
      }

      return "\(server.name) - "
    } catch {
      return ""
    }
  }

  private func processProgram(updatePrograms: [Program]) {
    for program in updatePrograms {
      do {
        if let exist = programs.firstIndex(where: { $0.program.serverId == program.serverId && $0.program.programKey == program.programKey }) {
          programs[exist] = try ProgramModel(program: program)
        } else {
          programs.append(try ProgramModel(program: program))
        }
      } catch {}
    }
  }
}
